### PR TITLE
Jenkins build module

### PR DESCRIPTION
**Pull request by brettmilford**
A module for queuing and deleting jenkins builds.

##### SUMMARY
Designed to have a similar interface as jenkins_job

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
jenkins_build

### CODE REVIEW

**Comment by felixfontein** [comment]
You need to create a symlink from `plugins/modules` to `plugins/modules/web_infrastructure/jenkins_build.py`.

File: plugins/modules/web_infrastructure/jenkins_build.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
+#!/usr/bin/python
+#
+# Copyright: (c) Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
```
Comment (felixfontein) [inline plugins/modules/web_infrastructure/jenkins_build.py:9]: This needs to be removed.

File: plugins/modules/web_infrastructure/jenkins_build.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: jenkins_build
+short_description: Manage jenkins builds
```
Comment (felixfontein) [inline plugins/modules/web_infrastructure/jenkins_build.py:16]:
```suggestion
short_description: Manage jenkins builds
version_added: 1.1.0
```

**Comment by brettmilford** [comment]
@felixfontein how's this?

**Comment by felixfontein** [comment]
Looks good to me. I don't know jenkins though, so no idea about the module itself :)

@stpierre @sermilrod @jtyr @hogarthj you are authors of other jenkins modules. Can you take a look?

File: plugins/modules/web_infrastructure/jenkins_build.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: jenkins_build
+short_description: Manage jenkins builds
+version_added: 1.1.0
```
Comment (Andersson007) [inline plugins/modules/web_infrastructure/jenkins_build.py:13]:
```suggestion
version_added: 1.3.0
```

File: plugins/modules/web_infrastructure/jenkins_build.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+description:
+    - Manage Jenkins builds with Jenkins REST API.
+requirements:
+  - "python-jenkins >= 0.4.12"
+author: "Brett Milford (@brettmilford)"
```
Comment (Andersson007) [inline plugins/modules/web_infrastructure/jenkins_build.py:18]:
```suggestion
author: Brett Milford (@brettmilford)
```

File: plugins/modules/web_infrastructure/jenkins_build.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+options:
+  args:
+    description:
+      - A list of parameters to pass to the build.
+    required: false
```
Comment (Andersson007) [inline plugins/modules/web_infrastructure/jenkins_build.py:23]:
```suggestion
```
this is a default. Please remove this from everywhere

File: plugins/modules/web_infrastructure/jenkins_build.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+url:
+  description: Url to connect to the Jenkins server.
+  returned: success
+  type: str
+  sample: https://jenkins.mydomain.com
+build_info:
+  description: Build info of the jenkins job.
+  returned: success
+  type: dict
+'''
```
Comment (Andersson007) [inline plugins/modules/web_infrastructure/jenkins_build.py:103]:
```suggestion
  type: dict
  sample: ...
'''
```

File: plugins/modules/web_infrastructure/jenkins_build.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+            password=dict(required=False, no_log=True),
+            state=dict(required=False, choices=['present', 'absent'], default="present"),
+            token=dict(required=False, no_log=True),
+            url=dict(required=False, default="http://localhost:8080"),
+            user=dict(required=False)
+        ),
+        mutually_exclusive=[
+            ['password', 'token'],
+        ],
+        supports_check_mode=True,
```
Comment (Andersson007) [inline plugins/modules/web_infrastructure/jenkins_build.py:234]:
```suggestion
        supports_check_mode=True,
```
if this is really supports `check_mode` (i don't see it's implementation in the code), please put it in the doc, e.g.:
```
note:
  - Supports C(check_mode).
```

**Review by Andersson007 — COMMENTED** [review]
@brettmilford thanks for the module! Generally looks good to me.

@felixfontein there's no feedback from other folks. I think we could merge the PR after my notes are fixed

File: plugins/modules/web_infrastructure/jenkins_build.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    module = AnsibleModule(
+        argument_spec=dict(
+            args=dict(required=False, type='dict'),
+            build_number=dict(required=False, type='int'),
+            name=dict(required=True),
+            password=dict(required=False, no_log=True),
+            state=dict(required=False, choices=['present', 'absent'], default="present"),
+            token=dict(required=False, no_log=True),
+            url=dict(required=False, default="http://localhost:8080"),
+            user=dict(required=False)
```
Comment (Andersson007) [inline plugins/modules/web_infrastructure/jenkins_build.py:229]:
```suggestion
            user=dict(required=False),
```
just a good practice:)

File: plugins/modules/web_infrastructure/jenkins_build.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+            if (self.user and self.password):
+                return jenkins.Jenkins(self.jenkins_url, self.user, self.password)
+            elif (self.user and self.token):
+                return jenkins.Jenkins(self.jenkins_url, self.user, self.token)
+            elif (self.user and not (self.password or self.token)):
+                return jenkins.Jenkins(self.jenkins_url, self.user)
+            else:
+                return jenkins.Jenkins(self.jenkins_url)
+        except Exception as e:
+            self.module.fail_json(msg='Unable to connect to Jenkins server, %s' % to_native(e), exception=traceback.format_exc())
```
Comment (Andersson007) [inline plugins/modules/web_infrastructure/jenkins_build.py:156]:
```suggestion
            self.module.fail_json(msg='Unable to connect to Jenkins server, %s' % to_native(e))
```
AFAIK, no needed any more

File: plugins/modules/web_infrastructure/jenkins_build.py (hunk 6 above)
Comment (brettmilford) [reply plugins/modules/web_infrastructure/jenkins_build.py:96]: I'd like to supply a sample for this but that's no longer feasible for me unfortunately.

**Comment by Andersson007** [comment]
jenkins folks @stpierre @jtyr @hogarthj could you please take a look at the new module?

File: plugins/modules/web_infrastructure/jenkins_build.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: jenkins_build
+short_description: Manage jenkins builds
+version_added: 1.3.0
```
Comment (felixfontein) [inline plugins/modules/web_infrastructure/jenkins_build.py:13]:
```suggestion
version_added: 2.2.0
```

**Review by felixfontein — APPROVED** [review]
Well since nobody from the jenkins folks reviewed this, I'll just merge this for the next release (2.2.0) if nobody complains until then (assuming CI still passes).

**Review by Andersson007 — APPROVED** [review]

**Comment by felixfontein** [comment]
@brettmilford thanks for contributing this!
@Andersson007 thanks for reviewing!

**Pull request merged**